File: cpp/problem234.cpp
#include <mystd.h>
using namespace MYSTD;
using namespace std;

//Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


/**
 * solution by reversing the link pointer of the second half of nodes
 *
 * There is a subtle memory leak. In principle we must repair the changed 
 * poitners in practical case, but we ignore this problem in this question 
 * for clarity.
 * 
 * Author: Yangyao, 2019/02/12
 */ 
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if( !head || !head->next ) return true;
        // find the middle point of the linked list
        ListNode *fast = head->next, *slow = head;
        while( fast && fast->next ){
            fast = fast->next->next;
            slow = slow->next;
        }
        // reverse the second half links
        auto *q = reverseLink( slow->next ), *p = head;
        // compare the first half and the second half
        while( q ){
            if( q->val != p->val ) return false;
            q = q->next;
            p = p->next;
        }
        return true;
    }
    ListNode *reverseLink(ListNode *head){
        ListNode *p = head->next, *q;
        head->next = nullptr;
        while(p){
            q = p->next;
            p->next = head;
            head = p;
            p = q;            
        }
        return head;
    }
};


int main(int argc, char const *argv[]){
    Solution sol;

    // test cases in Leetcode Problem
    {
        ListNode *head = new ListNode(1), *p = head;
        p->next = new ListNode(2); p = p->next;
        cout << sol.isPalindrome( head ) << endl;
    }
    {
        ListNode *head = new ListNode(1), *p = head;
        p->next = new ListNode(2); p = p->next;
        p->next = new ListNode(2); p = p->next;
        p->next = new ListNode(1); p = p->next;
        cout << sol.isPalindrome( head ) << endl;
    }
    return 0;
}

File: python/problem234.py
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        li = []
        while head:
            li.append(head.val)
            head = head.next
        
        l = 0
        r = len(li)-1
        while l<=r:
            if li[l] != li[r]:
                return False
            l += 1
            r -= 1
        return True
        
        
        
# O(1) space
# https://leetcode.com/problems/palindrome-linked-list/discuss/243254/super-clean-python-code
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        fast, slow, rev = head, head, None
        while fast and fast.next:
            fast = fast.next.next
            rev, rev.next, slow = slow, rev, slow.next
        if fast:
            slow = slow.next
        while rev and rev.val == slow.val:
            rev, slow = rev.next, slow.next
        return not rev